File: Creational_Patterns/Singleton/python/ThreadSafeSingleton.py
import threading

class ThreadSafeSingleton:
  _instance = None
  _lock = threading.Lock()

  def __init__(self):
    if ThreadSafeSingleton._instance is not None:
      raise Exception("There is a singleton instance")

  @classmethod
  def get_instance(cls):
    if cls._instance is None:
      with cls._lock:
        if cls._instance is None:
          cls._instance = ThreadSafeSingleton()
    return cls._instance

ob1 = ThreadSafeSingleton.get_instance()
ob2 = ThreadSafeSingleton.get_instance()

print(ob1 is ob2)

File: Creational_Patterns/Singleton/Java/ThreadSafeSingleton.java
class ThreadSafeSingleton{

  private static ThreadSafeSingleton instance = null;
  private ThreadSafeSingleton(){}

  public synchronized static ThreadSafeSingleton getInstance(){
    if(instance == null){
      instance = new ThreadSafeSingleton();
    }
    return instance;
  }
}

class Main{
  public static void main(String[] args){

    ThreadSafeSingleton ob1 = ThreadSafeSingleton.getInstance();
    ThreadSafeSingleton ob2 = ThreadSafeSingleton.getInstance();

    System.out.println(ob1 == ob2);
  }
}